### DISCUSSION TITLE
How to enable GPU for ocr_predictor?

### DISCUSSION THREAD

**Discussion by Chin-Shan-Hong** [mindee/doctr#1560] [General]
I am using "db_mobilenet_v3_large" and "crnn_mobilenet_v3_small" for OCR task. And I'm using CUDA 12.1 and NVIDIA GeForce RTX 3060. 

` model = ocr_predictor(det_arch='db_mobilenet_v3_large', reco_arch='crnn_mobilenet_v3_small',
                          pretrained=True, assume_straight_pages=True, det_bs=8, reco_bs=1024).cuda()`

I tried to add cuda() but have encountered AttributeError: 'OCRPredictor' object has no attribute 'cuda'

How can I enable GPU to speed up the OCR task?

**Comment by balochy** [comment]
you should wrap the whole predicator:

`model = (ocr_predictor(det_arch='db_mobilenet_v3_large', reco_arch='crnn_mobilenet_v3_small', pretrained=True, assume_straight_pages=True, det_bs=8, reco_bs=1024)).cuda()`

**Reply by Chin-Shan-Hong** [reply]
I still encounter the same error. Any ideas? I am using PyTorch for docTR

**Reply by balochy** [reply]
currently I use this configuration:

check if pytorch can detect cuda.
```
if torch.cuda.is_available():
    device = torch.device("cuda")
else:
    device = torch.device("cpu")

with torch.no_grad():
    model = (ocr_predictor(
        det_arch="db_resnet50",
        reco_arch=from_hub("Felix92/doctr-torch-parseq-multilingual-v1"),
        pretrained=True,
        straighten_pages=True,
        detect_orientation=True,
    ).to(device))
```

**Reply by Chin-Shan-Hong** [reply]
I tried to print(torch.cuda.is_available()) and it returns True. But I get AttributeError: 'OCRPredictor' object has no attribute 'to' when trying your method

**Reply by balochy** [reply]
import torch as well

**Reply by Chin-Shan-Hong** [reply]
Yes, I have imported torch too.

**Reply by balochy** [reply]
Hmm odd, wait for the project maintainer he can probably help you more

**Reply by Chin-Shan-Hong** [reply]
Ok, here is the snippet of my code

`import os
import torch
from doctr.io import DocumentFile
from doctr.models import ocr_predictor

os.environ['USE_TORCH'] = '1'
os.environ['USE_TF'] = '0'
os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"

 model = (ocr_predictor(det_arch='db_mobilenet_v3_large',
                               reco_arch='crnn_mobilenet_v3_small',
                                pretrained=True,
                               assume_straight_pages=True,
                               det_bs=8, reco_bs=1024,
                               detect_orientation=True).to(device))

AttributeError: 'OCRPredictor' object has no attribute 'to'
`

**Reply by felixdittrich92** [reply]
@Chin-Shan-Hong 
Seems like you have installed both TF and PT please run your script with `USE_TORCH=1 python3 /path/to/your_script.py` :)

```
predictor = ocr_predictor(...).cuda()  # .half()  if you want to run with float16- half precision
```

**Reply by Chin-Shan-Hong** [reply]
I don't quite understand this part `USE_TORCH=1 python3 /path/to/your_script.py`. I should add this to my script or I should execute it in my terminal?

**Reply by felixdittrich92** [reply]
You have 3 options:

First modify your snippet as follows:
```
import os

os.environ['USE_TF'] = '0'
os.environ['USE_TORCH'] = '1'
os.environ["KMP_DUPLICATE_LIB_OK"] = "TRUE"

import torch
from doctr.io import DocumentFile
from doctr.models import ocr_predictor

model = ocr_predictor(det_arch='db_mobilenet_v3_large', reco_arch='crnn_mobilenet_v3_small', pretrained=True, assume_straight_pages=True, det_bs=8, reco_bs=1024, detect_orientation=True).cuda()  # .half() for half precision
```

2:
Execute your script with `USE_TORCH=1` in front of the python command

3: (does the same as option 2)
```
export USE_TF=0
export USE_TORCH=1
```

**Reply by sanjay-nit** [reply]
Hi @felixdittrich92 

Setting above envs worked like a charm. I have few doubts:
1. Can we use ThreadPoolExecutor for list of images while using cuda ?
2. I used `model = ocr_predictor(pretrained=True, det_bs=8, reco_bs=1024).cuda()` i was continuously monitoring the colab gpu(T4) I feel like after processing hundreds of images sequentially few times; the GPU memory went from 3-4 Gbs to full memory and it again set back to 4 or 5 Gb.
3. When I used `model = ocr_predictor(pretrained=True).cuda()` the it was around 1.5Gbs throughout whole process(around 400 image process).

**Reply by LiquidGunay** [reply]
Yes I'm facing a similar issue that when I run ocr_predictor on a bunch of files in a sequence, my gpu memory usage spikes. It has crossed 24gb at one point causing a crash. Also, is there anyway to split this across multiple GPUs to avoid this kind of an error?

**Reply by sanjay-nit** [reply]
@LiquidGunay well, I don't think this has to be the solution to split this across multiple GPUs.
If you are processing an image at a time, You can avoid passing `det_bs` and `reco_bs` and it will pick default values which i guess is `2` and `128`. I used this while processing few hundreds of images sequentially and felt that GPU memory didn't cross 1.5 Gb.

**Reply by LiquidGunay** [reply]
@sanjay-nit I was running it with the default values as well, but I still ran into this issue of wildly fluctuating vram usage.

**Reply by sanjay-nit** [reply]
Interesting, thanks @LiquidGunay. you warned me to do not blindly trust and deploy it in production application 😄 😄 I was about to deploy it in my production app. Thanks for saving my life 😄 👍

**Reply by felixdittrich92** [reply]
Hi @LiquidGunay @sanjay-nit :wave:,

Please try the solution from https://github.com/mindee/doctr/discussions/1422
I agree with @sanjay-nit brute forcing with a sec GPU isn't a solution.

We use under the hood by default already threading so if you want to try something on top you should disable the internal usage (But i don't think that you will benefit from it.)

If you run on GPU you can also switch to half precision with `model = ocr_predictor(pretrained=True).cuda().half()` this will bring an speed up and lower the memory footprint.

You could also give https://github.com/felixdittrich92/OnnxTR a try it's more optimized for inference. (Note: this runs actually on my private account so the download rate limit is at 50GB i hope that we can move it to the mindee organization account to get rid of it or find another solution soon)

CC @odulcy-mindee wdyt ?

**Reply by sanjay-nit** [reply]
Hey @LiquidGunay I tried #1422 but didn't work for me.
below is more details on it https://github.com/mindee/doctr/issues/1594

**Reply by felixdittrich92** [reply]
@sanjay-nit Ah got it the discussion 1422 is about an increasing GPU VRAM usage so the flags doesn't change anything with RAM usage.
I saw that you process each image in a single.
Have you tried to pass it as a list ?
`doc = Documentfile.from_images([my_images])`

Otherwise a generator could help fetch & read your data if a batch is processed

Unfortunately i'm not able to reproduce this behaviour :/
I passed the whole CORD dataset as one (900 images) and processed with a det_bs of 4 and reco_bs of 1024 RAM hasn't hit 15GB ..

**Reply by sanjay-nit** [reply]
@felixdittrich92 Actually my use case is to deploy docTR server where it initialises and loads the model in memory and should ready for inference. So the model will be loaded only once and we send an image through api calls and it should do the ocr and return the result in response. I haven't tried deploying and end to end testing yet. but this seems like a problem because when I deploy the model it will be called sequentially so i don't see any solid reason it would work there.
Somehow it is not releasing the memory, i guess.

**Reply by sanjay-nit** [reply]
Any further suggestions? @felixdittrich92 #1594

**Reply by felixdittrich92** [reply]
Mh i tested it:

```
from doctr.io.reader import DocumentFile
from doctr.models import ocr_predictor
root_2 = "/home/felix/Desktop/Other/funsd_full/images/"  # 199 images
root = "/home/felix/Desktop/Other/cord_full/images/"  # 900 images

doctr_predictor = ocr_predictor(det_bs=2, reco_bs=1024, pretrained=True).cuda() #, load_in_8_bit=True)

files = [os.path.join(root, file) for file in os.listdir(root)] + [os.path.join(root_2, file) for file in os.listdir(root_2)]

for file in files:
    doc = DocumentFile.from_images([file])

    _ = doctr_predictor(doc)
   # proc sec time
    doc = DocumentFile.from_images([file])
    _ = doctr_predictor(doc)
```

this was constant <1.5 GB RAM usage .. :/

**Reply by felixdittrich92** [reply]
The highest peak i was able to reach was with passing complete CORD images at onces with ~41 GB RAM and then constant processing with ~36GB RAM  - but this should really only the case if you pass 100s of images at onces

**Reply by LiquidGunay** [reply]
My use case is similar, where I'm doing 100s of pdfs in sequence (10000+ images total). I haven't tested with the additional flags yet, only with half precision and default settings using dbresnet50 and parseq. At some point the vram usage crosses 22 gigs and my machine crashes.

**Reply by felixdittrich92** [reply]
ok @LiquidGunay about GPU VRAM the flags should help 

In the CPU RAM case @sanjay-nit you should test it (not sure what happened on Colab) but i double checked it and if you don't pass 100s of images at once to the DocumentFile the RAM usage was constant and the garbage collector did his job ^^

**Reply by sanjay-nit** [reply]
@felixdittrich92 Just FYI I tried the same what you suggested to pass the list of images rather than passing one by one. So after processing all the image it should have release the memory, right? And because it is not releasing the memory so what is happening let's say we have new documents/images coming up to process so again the memory uses will start from 36GB RAM in your case. I think it should not happen. You can see in below image after processing the 100 image it didn't release the memory so I process next 100 images so memory would increase by 2x. I hope you got my point.
![image](https://github.com/mindee/doctr/assets/103114249/b5f9f6da-f9cc-4344-8fe2-eeabe3e827f9)

**Reply by sanjay-nit** [reply]
@felixdittrich92 I tried the same thing in lambda GPU machine as well and there also I'm facing same issues.

**Reply by felixdittrich92** [reply]
@sanjay-nit Sure i got your point but unfortunately i can't reproduce it :sweat_smile:
